### DISCUSSION TITLE
How to catch and ignore errors in upstream dependency solids?

### DISCUSSION THREAD

**Discussion by liezl200** [dagster-io/dagster#3648] [Q&A]
My attempt didn't work:

```python
def create_process_solid(name: str, subject: str) -> solid:
    """
    Parameterize the _process_subject solid with subject.
    """

    @solid(name=name)
    def _process_subject(context: DagsterContext) -> str:
        context.log.info(subject)
        # process one subject
        with user_code_error_boundary(
            # Pass a class that inherits from DagsterUserCodeExecutionError
            DagsterExecutionStepExecutionError,
            # Pass a function that produces a message
            message_printer,
        ):
            if subject == "3":
                raise Exception("test exception")
        return f"successfully processed {subject}"

    return _process_subject
```

![Screen Shot 2021-02-03 at 6 39 14 PM](https://user-images.githubusercontent.com/1223842/106837034-2094bc00-664f-11eb-9584-f4c8694ad8fb.png)

Pipeline code

```
@pipeline
def process_pipeline() -> None:
    results = []
    subjects = ["1", "2", "3"]  # load from file
    for subject in subjects:
        # .../ingest/configs/covid.yaml --> covid.yaml --> covid_yaml
        process_subject = create_process_solid(f"process_subject_{subject}", subject)
        try:
            result = process_subject()
        except:
            result = "failed"
        results.append(result)

    # use done to force a dependency
    done = log_summary(results)

    # i guess we can't feed the input of process_subject into next_step_subject

    results2 = []
    for subject in subjects:
        next_step_subject = create_next_step_solid(
            f"next_step_subject_{subject}", subject, done
        )
        results2.append(next_step_subject(done))

    log_summary_next_step(results2)
```

**Comment by liezl200** [comment]
![Screen Shot 2021-02-03 at 6 41 48 PM](https://user-images.githubusercontent.com/1223842/106837552-90a34200-664f-11eb-895a-429953a9e947.png)

execution log 

desired behavior: pipeline continues but just shows a failed step

Ideally I wouldn't have to split and join after every parallel step but it seemed too hairy (and probably impossible?) to make a "compound solid" factory rather than multiple solid factories in sequence.

**Comment by liezl200** [comment]
@catherinewu @schrockn maybe?

**Comment by liezl200** [comment]
Hmm the easiest workaround would be to set up my own error reporting function handle_and_suppress_error to have full control over throwing and suppressing errors. That means that I would always have the main solid logic in a separate python_fn(args) and every solid would always call handle_and_suppress_error(python_fn(args))

Errors are rare enough that I would rather set up a small module to email me on failure rather than stop the entire pipeline.

Not an ideal solution because the UI wouldn't color the step red if it failed and was suppressed. Is there anyway to color the step red without halting execution?

**Reply by alangenfeld** [reply]
> Is there anyway to color the step red without halting execution?

Currently no, we don't allow any control over how failures are treated at this time. 

Since Dagster is data oriented in its graph dependencies, when a solid fails we consider its outputs invalid and there for can not pass them to downstream solids. If the behavior you seek is to continue execution by returning something else when an exception occurs, then the pattern you describe of catching errors in the solid body is probably the best thing you can do for now.

One other thing to note from your example above is that the `@pipeline` decorated function is only run at init time to capture the DAG structure. It is not used at all when the DAG is actually running.
